bundle/jsky.app.ot/src/main/scala/jsky/app/ot/gemini/editor/sitequality/TimingWindowImporter.scala

```diff
@@ -0,0 +1,213 @@
+}
+
+
+// Parser for TimingWindows.
+object TimingWindowParser extends RegexParsers {
+  // Whitespace is important, so do not skip it!
+  override val skipWhitespace = false
+
+  // Simple parsers.
+  // timeDigits must be a valid mm or ss field containing two digits.
+  private def whitespace: Parser[String] = """\s+""".r
+  private def arbitraryDigits: Parser[Int] = """\d+""".r ^^ { _.toInt }
+  private def timeDigits: Parser[Int] = """[0-5]\d""".r ^^ { _.toInt }
+
+  // A general parser to parse a temporal accessor from a DateTimeFormatter.
+  // NOTE: DateTimeFormatter spacing is RIGID and must be adhered to exactly.
+  private def temporalParser(df: DateTimeFormatter, failMsg: String): Parser[TemporalAccessor] = new Parser[TemporalAccessor] {
+    override def apply(in: Input): ParseResult[TemporalAccessor] = {
+      val source = in.source
+      val offset = in.offset
+      val start = handleWhiteSpace(source, offset)
+      val pos = new ParsePosition(start)
+      \/.fromTryCatchNonFatal {
+        val temporalAccessor = df.parse(source.toString, pos)
+        Success(temporalAccessor, in.drop(pos.getIndex - offset))
+      }.getOrElse(Failure(failMsg, in.drop(start - offset)))
+    }
+  }
+
+  // Parse a date and time in UTC into a Long.
+  // Formerly, in SimpleDateFormat, parsing would allow for arbitrary whitespace when a whitespace character was in the format.
+  // This is not the case with DateTimeFormatter parsing, so we must pre-process strings to change all whitespace sequences
+  // into single spaces.
+  private[sitequality] val dateTimeFormat = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss").withZone(ZoneId.of("UTC"))
+  private def dateParser: Parser[Long] =
+    temporalParser(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss").withZone(ZoneId.of("UTC")), "Failed to parse date") ^^ { case ta => Instant.from(ta).toEpochMilli }
+
+  // Convert hhhh:mm:ss as a duration to a Long in ms.
+  private def hhmmssToLong(hh: Int, mm: Int, ss: Int = 0): Long =
+    1000 * (60 * (60 * hh + mm) + ss)
+
+  // Convenience method for spec testing.
+  private[sitequality] def hhmmssStringToLong(hhmmss: String): Long =
+    hhmmss.split(':').toList.map(_.toInt) match {
+      case hh :: mm :: Nil       => hhmmssToLong(hh, mm)
+      case hh :: mm :: ss :: Nil => hhmmssToLong(hh, mm, ss)
+      case _                     => -1L
+    }
+
+  // Parse an hhhh:mm:ss duration / period and return the result in ms.
+  private def hhmmssParser: Parser[Long] = (arbitraryDigits <~ ":") ~ (timeDigits <~ ":") ~ timeDigits ^^ {
+    case hh ~ mm ~ ss => hhmmssToLong(hh, mm, ss)
+  }
+
+  private def hhmmParser: Parser[Long] = (arbitraryDigits <~ ":") ~ timeDigits ^^ {
+    case hh ~ mm => hhmmssToLong(hh, mm)
+  }
+
+  // Parse number of repetitions for a timing window:
+  // -1  = infinite repetitions.
+  //  0  = no repetitions.
+  // x>0 = x repetitions.
+  private def repetitionsParser: Parser[Int] = """(-1)|0|([1-9]\d*)""".r ^^ { _.toInt }
```

Comment:
Yes, Bryan said this was absolutely essential and listed -1 and 0 in his file format.